cmd/pullrequest-init/github.go

```diff
@@ -0,0 +1,227 @@
+	token := strings.TrimSpace(os.Getenv("GITHUBOAUTHTOKEN"))
+	var hc *http.Client
+	if token != "" {
+		ts := oauth2.StaticTokenSource(
+			&oauth2.Token{AccessToken: token},
+		)
+		hc = oauth2.NewClient(ctx, ts)
+	}
+
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return nil, err
+	}
+	split := strings.Split(u.Path, "/")
+	if len(split) != 5 {
+		return nil, fmt.Errorf("could not determine PR from URL: %v", rawURL)
+	}
+	owner, repo, pr := split[1], split[2], split[4]
+	prNumber, err := strconv.Atoi(pr)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing PR number: %s", pr)
+	}
+
+	return &GitHubHandler{
+		Client: github.NewClient(hc),
+		Logger: logger,
+		owner:  owner,
+		repo:   repo,
+		prNum:  prNumber,
+	}, nil
+}
+
+// writeJSON writes an arbitrary interface to the given path.
+func writeJSON(path string, i interface{}) error {
+	f, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	return json.NewEncoder(f).Encode(i)
+}
+
+// Download fetches and stores the desired pull request.
+func (h *GitHubHandler) Download(ctx context.Context, path string) error {
+	rawPrefix := filepath.Join(path, "github")
+	if err := os.MkdirAll(rawPrefix, 0755); err != nil {
+		return err
+	}
+
+	// Pull request
+	gpr, _, err := h.PullRequests.Get(ctx, h.owner, h.repo, h.prNum)
+	if err != nil {
+		return err
+	}
+	rawPR := filepath.Join(rawPrefix, "pr.json")
+	if err := writeJSON(rawPR, gpr); err != nil {
+		return err
+	}
+	pr := &PullRequest{
+		Type: "github",
+		ID:   gpr.GetID(),
+		Head: &GitReference{
+			Repo:   gpr.GetHead().GetRepo().GetCloneURL(),
+			Branch: gpr.GetHead().GetRef(),
+			SHA:    gpr.GetHead().GetSHA(),
+		},
+		Base: &GitReference{
+			Repo:   gpr.GetBase().GetRepo().GetCloneURL(),
+			Branch: gpr.GetBase().GetRef(),
+			SHA:    gpr.GetBase().GetSHA(),
+		},
+
+		Raw: rawPR,
+	}
+
+	// Labels
+	pr.Labels = make([]*Label, 0, len(gpr.Labels))
+	for _, l := range gpr.Labels {
+		pr.Labels = append(pr.Labels, &Label{
+			Text: l.GetName(),
+		})
+	}
+
+	// Comments
+	commentsPrefix := filepath.Join(rawPrefix, "comments")
+	for _, p := range []string{commentsPrefix} {
+		if err := os.MkdirAll(p, 0755); err != nil {
+			return err
+		}
+	}
+	ic, _, err := h.Issues.ListComments(ctx, h.owner, h.repo, h.prNum, nil)
+	if err != nil {
+		return err
+	}
+	pr.Comments = make([]*Comment, 0, len(ic))
+	for _, c := range ic {
+		rawComment := filepath.Join(commentsPrefix, fmt.Sprintf("%d.json", c.GetID()))
+		h.Logger.Infof("Writing comment %d to file: %s", c.GetID(), rawComment)
+		if err := writeJSON(rawComment, c); err != nil {
+			return err
+		}
+
+		comment := &Comment{
+			Author: c.GetUser().GetLogin(),
+			Text:   c.GetBody(),
+			ID:     c.GetID(),
+
+			Raw: rawComment,
+		}
+		pr.Comments = append(pr.Comments, comment)
+	}
+
+	prPath := filepath.Join(path, prFile)
+	h.Logger.Infof("Writing pull request to file: %s", prPath)
+	return writeJSON(prPath, pr)
```

Comment:
this function seems a bit long - could we break it up into several functions with unit tests? (e.g. one idea: sep. functions for labels, comments, getting the pull request)